tensorflow/lite/micro/kernels/xcore/xcore_dispatcher.cc

```diff
@@ -40,138 +46,76 @@ Dispatcher* GetDispatcher() {
 
 #ifdef XCORE
 
-// xCORE Dispatcher implementation.
-// Uses a threadgroup_t to dispatch tasks to threads.
-Dispatcher::Dispatcher(tflite::ErrorReporter* reporter, bool use_current_core)
-    : use_current_thread_(use_current_core), reporter_(reporter) {
-  group_ = thread_group_alloc();
-  tasks_.size = 0;
-}
-
-Dispatcher::~Dispatcher() { thread_group_free(group_); }
-
-TfLiteStatus Dispatcher::JoinTasks() {
-  if (tasks_.size == 0) return kTfLiteOk;
-
-  int begin = 0;
-
-  if (use_current_thread_) {
-    // reserves the first task to be spawned later in this thread
-    begin++;
-  }
-
-  int remaining_tasks = tasks_.size - begin;
-
-  if (remaining_tasks > 0) {
-    // add tasks to the thread group
-    size_t stack_words = tasks_.stack_size / kBytesPerStackword;
-
-    // align up the stack pointer (if necessary)
-    uint8_t* aligned_stack = AlignPointerUp(
-        reinterpret_cast<uint8_t*>(tasks_.stack), kDoubleWordAlignment);
-
-    for (int i = begin; i < tasks_.size; i++) {
-      thread_group_add(group_, tasks_.function, tasks_.arguments[i],
-                       stack_base(aligned_stack, stack_words));
-      aligned_stack = AlignPointerUp(&aligned_stack[tasks_.stack_size],
-                                     kDoubleWordAlignment);
-    }
-
-    // spawn the thread group
-    thread_group_start(group_);
-
-    if (use_current_thread_) {
-      // spawn the first task in this thread
-      (tasks_.function)(tasks_.arguments[0]);
+TfLiteStatus GenericDispatcher::Invoke(
+    const dispatcher_arg_array_t& arguments) const {
+  threadgroup_t group = thread_group_alloc();
+
+  size_t stack_words = stack_size_ / kBytesPerStackword;
+
+  // Align up the stack pointer (if necessary)
+  uint8_t* aligned_stack = AlignPointerUp(
+      reinterpret_cast<uint8_t*>(stack_memory_), kDoubleWordAlignment);
+
+  size_t num_threads_added = 0;
+  for (int i = 0; i < arguments.size(); i++) {
+    thread_group_add(group, function_, arguments[i],
+                     stack_base(aligned_stack, stack_words));
+
+    aligned_stack =
+        AlignPointerUp(&aligned_stack[stack_size_], kDoubleWordAlignment);
```

Comment:
Looks like this could increment `aligned_stack` too far if `arguments.size()` is greater than `num_threads_`.